craft_application/services/service_factory.py

```diff
@@ -71,24 +71,25 @@ def __getattr__(self, service: str) -> services.BaseService:
         instantiated service as an instance attribute, allowing the same service
         instance to be reused for the entire run of the application.
         """
-        if self.project is None:  # pyright: ignore[reportUnnecessaryComparison]
-            raise errors.ApplicationError(
-                "ServiceFactory requires a project to be set before getting a service.",
-                app_name=self.app.name,
-                docs_url="https://github.com/canonical/craft-application/pull/40#discussion_r1253593262",
-            )
         service_cls_name = "".join(word.title() for word in service.split("_"))
         service_cls_name += "Class"
         classes = dataclasses.asdict(self)
         if service_cls_name not in classes:
             raise AttributeError(service)
         cls = getattr(self, service_cls_name)
-        if issubclass(cls, services.BaseService):
-            kwargs = self._service_kwargs.get(service, {})
-            instance = cls(self.app, self.project, self, **kwargs)
-            instance.setup()
-            setattr(self, service, instance)
-            # Mypy and pyright interpret this differently.
-            # Pyright
-            return instance  # type: ignore[no-any-return]
-        raise TypeError(f"{cls.__name__} is not a service class")
+        if not issubclass(cls, services.AppService):
+            raise TypeError(f"{cls.__name__} is not a service class")
+        kwargs = self._service_kwargs.get(service, {})
+        if issubclass(cls, services.ProjectService):
+            if not self.project:
+                raise errors.ApplicationError(
```

Comment:
the logic we have is for _native_ errors to pop up as an internal error and craft exceptions to show up as a user friendly thing, this one seems like a developer error for which we do not want a end user type of handling for.

Reply:
Yeah, the original idea for `ApplicationError` was to provide nice error messages when in debug mode, but as the framework has evolved it's been getting used less and less. I've updated it to a `ValueError` and removed `ApplicationError` since this was the last remaining instance.